repository/student/student.go

```diff
@@ -0,0 +1,157 @@
+	ErrAlreadyExists = errors.New("student already exists")
+)
+
+// Student represents a student record in the database.
+type Student struct {
+	ID        string    `db:"id"`
+	FirstName string    `db:"first_name"`
+	LastName  string    `db:"last_name"`
+	Grade     int32     `db:"grade"`
+	CreatedAt time.Time `db:"created_at"`
+}
+
+// Repository manages operations with students in the DB.
+type Repository struct {
+	db *sqlx.DB
+}
+
+// NewRepository creates a new Repository with the given DB connection.
+func NewRepository(db *sqlx.DB) *Repository {
+	return &Repository{db: db}
+}
+
+// Create inserts a new student record into the database.
+func (r *Repository) Create(ctx context.Context, s Student) (string, error) {
+	tx, err := r.db.BeginTxx(ctx, nil)
+	if err != nil {
+		return "", err
+	}
+
+	query := fmt.Sprintf(`INSERT INTO %s (%s) VALUES %s`, tableName, ColumnsStr(), NamedPlaceholders())
+
+	_, err = tx.NamedExecContext(ctx, query, &s)
+
+	if err != nil {
+		if rbErr := tx.Rollback(); rbErr != nil {
+			log.Printf("tx.Rollback error: %v", rbErr)
+		}
+		return "", err
+	}
+
+	if err := tx.Commit(); err != nil {
+		return "", err
+	}
+
+	return s.ID, nil
+}
+
+// GetByID retrieves a student by their ID.
+func (r *Repository) GetByID(ctx context.Context, id string) (*Student, error) {
+	var s Student
+	query := fmt.Sprintf(`SELECT id, first_name, last_name, grade, created_at FROM %s WHERE id = $1`, tableName)
+	err := r.db.GetContext(ctx, &s, query, id)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, ErrNotFound
+		}
+		return nil, err
+	}
+	return &s, nil
+}
+
+// Update modifies an existing student record.
+func (r *Repository) Update(ctx context.Context, s Student) error {
+	tx, err := r.db.BeginTxx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			_ = tx.Rollback()
+		} else {
+			err = tx.Commit()
+		}
+	}()
+
+	query := fmt.Sprintf(`UPDATE %s SET %s WHERE id = :id`, tableName, UpdateSetStr())
+
+	result, err := tx.NamedExecContext(ctx, query, &s)
+	if err != nil {
+		return err
+	}
+
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	if rowsAffected == 0 {
+		return ErrNotFound
+	}
+
+	return nil
+}
+
+// Delete removes a student record by ID.
+func (r *Repository) Delete(ctx context.Context, id string) error {
+	tx, err := r.db.BeginTxx(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	query := fmt.Sprintf(`DELETE FROM %s WHERE id = $1`, tableName)
+
+	result, err := tx.ExecContext(ctx, query, id)
+	if err != nil {
+		if rbErr := tx.Rollback(); rbErr != nil {
+			log.Printf("tx.Rollback error: %v", rbErr)
+		}
+		return err
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		if rbErr := tx.Rollback(); rbErr != nil {
+			log.Printf("tx.Rollback error: %v", rbErr)
+		}
+		return err
+	}
+	if rowsAffected == 0 {
+		if rbErr := tx.Rollback(); rbErr != nil {
```

Comment:
Why rollback in this case? Commit!

Reply:
Otherwise concurrent create request could complete

Reply:
> doesn't finished successfully
It has finished successfully in terms of performing delete operation - all record are successfully deleted. The fact that there are no records is not an error in terms of performing database request. It could be considered to be error on repository layer, since it is not intended to be used this way, but it is still a successful transaction.
In other words, delete caller, that received NotFound, would expect guarantees on absence of the object -- transactional guarantees